pandas/core/groupby.py

```diff
@@ -2314,10 +2341,13 @@ def _cython_operation(self, kind, values, how, axis, min_count=-1):
             else:
                 raise
 
-        if is_numeric:
-            out_dtype = '%s%d' % (values.dtype.kind, values.dtype.itemsize)
+        if how == 'rank':
```

Comment:
we should prob have a better way of doing this :<

Reply:
The catch I see here is that unlike other groupby transformations (where the dtype of the `result` object typically matches the dtype of the groups' values and gets cast as necessary after transformation) the dtype of the result object for rank needs to be a float prior to calling any `group_rank` operation. Otherwise, things like TIEBREAK_AVERAGE will not work when ranking say a group full of ints.

The generic calls to `algos` have it a little bit easier because they don't pass an object in to be modified, using instead the return value which `algos` builds internally as a like-sized array of floats. Unless there's something basic I don't understand with Cython, I don't think there's a way to upcast the dtype of the `return` object that gets passed into `group_rank` from say an int to a float, so the only option I can think of there would be to break out `group_rank` from other transformations and have `group_rank` return a new `result` object instead of modifying the provided one in place.

I'd argue the effort and maintainability of doing that would far outweigh any benefit from cleaning up this conditional, but curious if you have other thoughts

Reply:
this is similar in a way to the pre-transformations we do for datetimelikes. What I meant is that the registry of _cython_functions should have this logic itself (which is already based on the function name). 

e.g. maybe in _cython_functions you get back a tuple rather than a scalar / lamba function, which includes the required dtype (or None).  Or _cython_functions actually should be a class which is dispatched to, to call the functions themselves. IOW something like

```
class CythonTransformFunction:
    def __init__(self, obj):
        self.obj = obj

    def cummax(self):
         return group_cummax(self.obj.values)

    def rank(self):
        return group_cummax(self.obj.values.astype('floatt64'))
```

(and really should have more logic pushed to this class, e.g. pre-and post-convert dtype things. And should have a Transformation and an Aggregation version. This would all of the specific logic to be pushed here, rather than lookup in dicts and such. Further we could prob move this code out of the main groupby.py class into a sub-module.

don't have to do this here, but groupby for sure needs cleanup like this.